From c75aa7e75aa405cbc4cd49de0351aae757c7dbb0 Mon Sep 17 00:00:00 2001
From: AhmedAdelAbouzeid <ahmed.adel.abouzeid@gmail.com>
Date: Fri, 17 Apr 2020 12:16:30 +0200
Subject: [PATCH 1/2] Correct on-screen keyboard behaviour for Arabic
 characters 1- Prevent arabic characters from being disjoined when using the
 on-screen keyboard. Fixes #16040 2- Prevent the flipping of existing arabic
 text when editing (ex. renaming something) using the on-screen keyboard

---
 xbmc/guilib/GUIEditControl.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xbmc/guilib/GUIEditControl.cpp b/xbmc/guilib/GUIEditControl.cpp
index e04c96ff3fa8..ded5f3ecdf87 100644
--- a/xbmc/guilib/GUIEditControl.cpp
+++ b/xbmc/guilib/GUIEditControl.cpp
@@ -261,7 +261,7 @@ bool CGUIEditControl::OnAction(const CAction &action)
     {
       m_edit.clear();
       std::wstring str;
-      g_charsetConverter.utf8ToW(action.GetText(), str);
+      g_charsetConverter.utf8ToW(action.GetText(), str, false);
       m_text2.insert(m_cursorPos, str);
       m_cursorPos += str.size();
       UpdateText();
@@ -587,7 +587,7 @@ void CGUIEditControl::SetLabel2(const std::string &text)
 {
   m_edit.clear();
   std::wstring newText;
-  g_charsetConverter.utf8ToW(text, newText);
+  g_charsetConverter.utf8ToW(text, newText, false);
   if (newText != m_text2)
   {
     m_isMD5 = (m_inputType == INPUT_TYPE_PASSWORD_MD5 || m_inputType == INPUT_TYPE_PASSWORD_NUMBER_VERIFY_NEW);

From e997c87d28c0b9fe4e2da6fa04a52f70e5191da0 Mon Sep 17 00:00:00 2001
From: AhmedAdelAbouzeid <ahmed.adel.abouzeid@gmail.com>
Date: Tue, 21 Apr 2020 09:41:01 +0200
Subject: [PATCH 2/2] Adjust cursor blinking effect for on-screen keyboard to
 accommodate Arabic text

---
 xbmc/guilib/GUIEditControl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/guilib/GUIEditControl.cpp b/xbmc/guilib/GUIEditControl.cpp
index ded5f3ecdf87..1c6d9870ac06 100644
--- a/xbmc/guilib/GUIEditControl.cpp
+++ b/xbmc/guilib/GUIEditControl.cpp
@@ -565,7 +565,7 @@ bool CGUIEditControl::SetStyledText(const std::wstring &text)
   // show the cursor
   unsigned int ch = L'|';
   if ((++m_cursorBlink % 64) > 32)
-    ch |= (3 << 16);
+    ch = L' ';
   styled.insert(styled.begin() + m_cursorPos, ch);
 
   return m_label2.SetStyledText(styled, colors);
